### ISSUE TITLE
Kmeans GPU

### ISSUE THREAD

**Issue by conceptofmind** [facebookresearch/faiss#2871]
# Summary

Hi all,

I was wondering what the correct way would be to enable GPU acceleration for Kmeans clustering using a node with 8 GPUs? I previously indexed on the CPU and cloned to the GPU but Kmeans has an optional flag for `gpu=True`.

Running on:
- [X] CPU
- [X] GPU

Interface: 
- [ ] C++
- [X] Python

# Reproduction instructions

Currently, the code I have is:
```python
# Normalize the embeddings
embeddings = np.array(embeddings).astype("float32")  # FAISS uses float32
embeddings = normalize(embeddings)

# perform clustering with FAISS
num_clusters = 11000
niter = 20
epsilon = 0.09  # Define the similarity threshold
verbose = True
d = embeddings.shape[1]  # dimension

# Initialize the clustering
cpu_index = faiss.Kmeans(d, num_clusters, niter=niter, verbose=verbose, spherical=True)
cpu_index.train(embeddings)

# Number of GPU resources
ngpus = faiss.get_num_gpus()
res = [faiss.StandardGpuResources() for _ in range(ngpus)]

# Cloner options
co = faiss.GpuMultipleClonerOptions()
co.shard = True

# Initialize the index. clone index to GPUs
index = faiss.index_cpu_to_all_gpus(cpu_index, co=co, resources=res)

D, I = index.search(embeddings, 1)
cluster_labels = I.reshape(-1)
cluster_centers = index.centroids
```
Should it instead be:
```python

# Normalize the embeddings
embeddings = np.array(embeddings).astype("float32")  # FAISS uses float32
embeddings = normalize(embeddings)

# perform clustering with FAISS
num_clusters = 11000
niter = 20
epsilon = 0.09  # Define the similarity threshold
verbose = True
d = embeddings.shape[1]  # dimension

# Initialize the clustering
kmeans = faiss.Kmeans(d, num_clusters, niter=niter, verbose=verbose, spherical=True, gpu=True)
kmeans.train(embeddings)

D, I = kmeans.index.search(embeddings, 1)
cluster_labels = I.reshape(-1)
cluster_centers = kmeans.centroids

points_to_keep = []

for i in range(num_clusters):
    # filter embeddings of the current cluster
    cluster_i_embeddings = embeddings[cluster_labels == i]

    # sort the cluster embeddings by the distance to the cluster centroid
    cluster_i_embeddings = sort_by_centroid_distance(
        cluster_i_embeddings, cluster_centers[i], descending=True
    )

    # compute the pairwise cosine similarity between embeddings
    pairwise_sim_matrix = cosine_similarity(cluster_i_embeddings)

    # get upper triangular part of the matrix (excluding the diagonal)
    triu_sim_matrix = np.triu(pairwise_sim_matrix, k=1)

    # find max value in each column
    M = np.max(triu_sim_matrix, axis=0)[0]

    # Check if the maximum similarity <= the threshold.
    points_to_keep_from_cluster_i = cluster_i_embeddings[M <= 1 - epsilon]

    # add the points to keep to the list
    points_to_keep.extend(points_to_keep_from_cluster_i)
```
Any help would be greatly appreciated!

Thank you,

Enrico

**Comment by mdouze** [comment]
see
https://github.com/facebookresearch/faiss/blob/1c1879b17c6a7030a4cc7a88535c66cf38cdf4c9/faiss/python/extra_wrappers.py#L386

**Comment by conceptofmind** [comment]
> see
> 
> https://github.com/facebookresearch/faiss/blob/1c1879b17c6a7030a4cc7a88535c66cf38cdf4c9/faiss/python/extra_wrappers.py#L386

Hi @mdouze ,

Thank you for the reference.

I see that the GPU version of Kmeans handles: 
1. The conversion of the dtype to fp32 - `np.array(embeddings).astype("float32")`
2. The number of GPUs available in the environment - `ngpus = faiss.get_num_gpus() res = [faiss.StandardGpuResources() for _ in range(ngpus)]`
3. The placement of the indices onto the GPU - `faiss.index_cpu_to_all_gpus(cpu_index, co=co, resources=res)`

I would think this also automatically handles the cloner - `co = faiss.GpuMultipleClonerOptions() co.shard = True`

I was wondering if Spherical is set to `True` whether the embeddings would still need to be normalized before? Or does Spherical already handle the normalization process?

For example, the normalization step would no longer be needed:
```python
from sklearn.preprocessing import normalize
embeddings = normalize(embeddings)
```
I greatly appreciate all of your help.

Thank you,

Enrico
